src/transforms/util/pick.rs

```diff
@@ -0,0 +1,121 @@
+//! Dynamically pick a tranform to use.
+
+#![deny(missing_docs)]
+
+use crate::{event::Event, transforms::Transform};
+
+/// Dynamically pick a tranform to use.
+///
+/// The implementation would have multiple candidates and a way to determine
+/// which transform to use.
+pub trait Picker: Sized {
+    /// Consumes an event and the picker itself, and returns a transform to
+    /// use and a transformed version of the passed event.
+    ///
+    /// Note that the interface doesn't allow failures (i.e. it's not a
+    /// [`Result`], nor [`Option`]).
+    /// To handle failure modes, use provided [`Passthrough`], [`Discard`] or
+    /// [`Panic`] as the returned transform, and pass the event accordingly.
+    fn probe_event(self, event: Event) -> (Box<dyn Transform>, Option<Event>);
+}
+
+/// Dynamically pick a transform to use once (on first event), and then use the
+/// picked transform for the rest of the events.
+///
+/// The first message obtained will be probed with picker, and then the system
+/// will switch to using.
+pub enum PickOnce<T>
+where
+    T: Picker + Send,
+{
+    /// The initial state.
+    Init(Option<T>),
+    /// The picked transform.
+    Picked(Box<dyn Transform>),
+}
+
+impl<T> PickOnce<T>
+where
+    T: Picker + Send,
+{
+    /// Create a new [`PickOnce`] using the passed `T` as the initial state.
+    pub fn new(init: T) -> Self {
+        Self::Init(Some(init))
+    }
+}
+
+impl<T> Transform for PickOnce<T>
+where
+    T: Picker + Send,
+{
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        match self {
+            Self::Init(picker) => {
+                let picker = picker.take().expect("impossible state");
+                let (transform, event) = picker.probe_event(event);
+                *self = Self::Picked(transform);
+                event
+            }
+            Self::Picked(transform) => transform.transform(event),
+        }
+    }
+}
+
+/// A passthrough transform. Passes events as-is.
+pub struct Passthrough;
+
+impl Transform for Passthrough {
+    fn transform(&mut self, event: Event) -> Option<Event> {
+        Some(event)
+    }
+}
+
+/// A discard transform. Returns `None` for every event.
+pub struct Discard;
```

Comment:
Not sure this is used?

Reply:
Not used, you're right.

Reply:
We can safely remove those. Though, this is a by-design-should-be-there kind of thing. Cutting it out would make the design incomplete, not sure if it makes sense. You'd typically have the core primitives unsued when you implement composition frameworks, so I think this is normal. Someone might find it useful later on.

Not a big fan of unused stuff lying around the code too though, but this seems like something we can make an exception for.

Reply:
So how is the design incomplete if you can remove them an nothing breaks?

Reply:
Yeah, that all makes sense :) Just not sure where it's actually used? Testing?

Reply:
Eeh, I'm not sure I get what you're asking :frowning: Currently, nothing uses this code.
But I'm hoping it's there in the future if/when I'm going to need it. I can't predict whether it'll actually happen, so, currently, this is a dead code.
This is not a usual thing to have dead code that can be used later though, is it? For instance, we're not using the whole API provided by the `serde` or `tokio` crates. They just implement things that will be useful for people.
The approach here is very similar. We have a unit that has a certain API, is designed to be a reusable piece, and is not in itself a direct feature implementation (it is a transform, but it's not usable via topology). I'd suggest that it's ok to apply the same rules we apply to libraries to the utility code like this.
I hope what I wrote about is relevant, if not - let's quickly exchange messages in DMs.

Reply:
So I had some similar cases in the WASM work (eg I needed to make sure I could support panics, so I made a panic module to test with), and I made them have a unit test to make sure it's not just totally dead code.

Reply:
I see what you mean. Yes, tests are mandatory here. I'm going to do another pass with adding tests. I didn't want to spend time on tests, cause I wanted to feedback first.
I expected the whole `transform::util` changes to be a hot topic.

Reply:
> Currently, nothing uses this code.

Then the code should not exist. That is the end of the story. Every line of code has a cost and there is no justification for paying the cost of code that is entirely unused.

> For instance, we're not using the whole API provided by the `serde` or `tokio` crates.

There is no comparison here. We are not the only users of those APIs. If there is even one small part of those APIs that are not at all used by anyone anywhere, they absolutely should be deleted as well.

> But I'm hoping it's there in the future if/when I'm going to need it. I can't predict whether it'll actually happen

If it happens that you do need it, write the code then! Unless you believe the code will somehow take you longer to write in the future, you've almost certainly wasted time by writing it now.

Reply:
Yeah, well, you're right Idk why I was arguing this yesterday, bad day.
I'll move things around so they're in the `kubernetes_logs` and private. Or does it make sense to keep them share, and just cut the unused bits?

Reply:
I cleaned up the code, please let me know if you think this is good enough.